packages/validate/src/validate.ts

```diff
@@ -485,18 +485,23 @@ function impUnion<
   T extends [Validator<unknown>, Validator<unknown>, ...Validator<unknown>[]],
 >(elements: T, clone: boolean) {
   return function impUnion(value: unknown): Result<Infer<T[number]>> {
-    for (const element of elements) {
+    const errors: string[] = []
+    for (const [i, element] of elements.entries()) {
       const result = clone
         ? element.safeParse(value)
         : element.safeValidate(value)
       if (result.success) {
         return result as Result<Infer<T[number]>>
       }
+
+      const location = result.path ? ` at ${result.path}` : ''
+      errors.push(`Variant ${i}${location}: ${result.message}`)
     }
+    const errorDetails = errors.length > 0 ? ` ${errors.join(' ')}` : ''
```

Comment:
Instead of this and the error array and handling the space in front why not just have a string message and say `message += ` Variant ${i}${location}: ${result.message}`. That way we don't need to join and worry about the space in front and check errors.length.